langchain/agents/executor.ts

```diff
@@ -74,9 +84,33 @@ export class AgentExecutor extends BaseChain {
       }
 
       const tool = toolsByName[action.tool.toLowerCase()];
-      const observation = tool
-        ? await tool.call(action.toolInput)
-        : `${action.tool} is not a valid tool, try another one.`;
+      this.callbackManager(
+        {
+          event: "agent.tool_start",
+          tool: { name: tool?.name ?? "N/A" },
+          action,
+        },
+        { verbose: this.verbose }
+      );
+
+      let observation;
+      try {
+        observation = tool
+          ? await tool.call(action.toolInput)
+          : `${action.tool} is not a valid tool, try another one.`;
+      } catch (err) {
+        this.callbackManager(
+          { event: "agent.tool_error", err },
+          { verbose: this.verbose }
+        );
+        throw err;
+      }
+
+      this.callbackManager(
```

Comment:
With async I meant even this method should be async, otherwise we end up dangling promises